Tenant ID propagation sample (#4256)
Particular/docs.particular.net: samples/multi-tenant/propagation/sample.md

---
title: Propagating tenant information to downstream endpoints
summary: Shows how to configure the pipeline to automatically propagate tenant information to downstream endpoints
reviewed: 2019-02-08
component: Core
related:
- nservicebus/pipeline/manipulate-with-behaviors
---

## Introduction

> This sample shows how to configure the NServiceBus pipeline to automatically propagate tenant information to the downstream endpoints. The sample assumes that the tenant information is passed as a message header.

Reviewer: ```suggestion
This sample shows how to configure the NServiceBus pipeline to automatically propagate tenant information to downstream endpoints. The sample assumes that the tenant information is passed as a message header.
```

## Code Walk Through

> ### Attaching the tenant information to the message

Reviewer: ```suggestion
### Attaching tenant information to the messages
```

> In most cases the best way to attach tenant information message is by using a custom message header. Following code demonstrates how to set a custom header `tenant_id`.

Reviewer: ```suggestion
In most cases the best way to attach tenant information to messages is by using a custom message header. The following code demonstrates how to set a custom `tenant_id` header.
```

snippet: SetTenantId

> ### Creating the Behaviors

Reviewer: ```suggestion
### Creating Behaviors
```

> There are two behaviors required to propagate the tenant information.

Reviewer: ```suggestion
Two behaviors are required to propagate the tenant information.
```

#### Retrieving and storing the tenant information

> The first behavior is responsible for extracting the tenant information from a message header and placing it in the pipeline execution context bag. This behavior executes as part of the message receive pipeline.

Reviewer: ```suggestion
The first behavior is responsible for extracting the tenant information from the incoming message header, and placing it in the pipeline execution context bag. This behavior executes as part of the message receive pipeline.
```

snippet: StoreTenantId

#### Propagating the tenant information to the outgoing messages

> The second behavior is responsible for attaching the tenant information header(s) to the outgoing messages based on the pipeline context. This behavior executes as part of the message send pipeline.

Reviewer: ```suggestion
The second behavior is responsible for attaching the tenant information header(s) to outgoing messages based on the pipeline context. This behavior executes as part of the message send pipeline.
```

snippet: PropagateTenantId

### Registering the behaviors

> The following code is needed to register the behavior in the pipeline.

Reviewer: ```suggestion
The following code is needed to register the created behaviors in the pipeline.
```
